timemachine/fe/single_topology_v3.py

```diff
@@ -0,0 +1,539 @@
+
+        for a, b in core:
+            self.c_flags[a] = 0
+            self.b_to_c[b] = a
+
+        iota = self.mol_a.GetNumAtoms()
+        for b_idx, c_idx in enumerate(self.b_to_c):
+            if c_idx == -1:
+                self.b_to_c[b_idx] = iota
+                self.c_flags[iota] = 2
+                iota += 1
+
+        # test for uniqueness in core idxs for each mol
+        assert len(set(tuple(core[:, 0]))) == len(core[:, 0])
+        assert len(set(tuple(core[:, 1]))) == len(core[:, 1])
+
+    def get_num_atoms(self):
+        """
+        Get the total number of atoms in the alchemical hybrid.
+
+        Returns
+        -------
+        int
+            Total number of atoms.
+        """
+        return self.NC
+
+    def combine_confs(self, x_a, x_b):
+        """
+        Combine conformations of two molecules
+
+        Parameters
+        ----------
+        x_a: np.array of shape (N_A,3)
+            First conformation
+
+        x_b: np.array of shape (N_B,3)
+            Second conformation
+
+        Returns
+        -------
+        np.array of shape (self.num_atoms,3)
+            Combined conformation
+
+        """
+        assert x_a.shape == (self.mol_a.GetNumAtoms(), 3)
+        assert x_b.shape == (self.mol_b.GetNumAtoms(), 3)
+        x0 = np.zeros((self.get_num_atoms(), 3))
+        for src, dst in enumerate(self.a_to_c):
+            x0[dst] = x_a[src]
+        for src, dst in enumerate(self.b_to_c):
+            x0[dst] = x_b[src]
+        return x0
+
+    def setup_end_state_src(self):
+        """
+        Setup the source end-state, where mol_a is fully interacting, with mol_b's dummy atoms attached
+        in a factorizable way.
+
+        Returns
+        -------
+        VacuumSystem
+            Gas-phase system
+        """
+        return setup_end_state(self.ff, self.mol_a, self.mol_b, self.core, self.a_to_c, self.b_to_c)
+
+    def setup_end_state_dst(self):
+        """
+        Setup the source end-state, where mol_a is fully interacting, with mol_b's dummy atoms attached
+        in a factorizable way.
+
+        Returns
+        -------
+        VacuumSystem
+            Gas-phase system
+        """
+        new_core = np.stack([self.core[:, 1], self.core[:, 0]], axis=1)
+        return setup_end_state(self.ff, self.mol_b, self.mol_a, new_core, self.b_to_c, self.a_to_c)
+
+    def get_U_fn(self, lamb):
+        """
+        Get a jax compatible energy function parameterized at some value of lambda, using linear
+        energy interpolation.
+
+        Parameters
+        ----------
+        lamb: float
+            0 <= lamb <= 1
+
+        Returns
+        -------
+        Callable:
+            An energy function f: R^(Nx3) -> R^1
+
+        """
+        U0_fn = self.setup_end_state_src().get_U_fn()
+        U1_fn = self.setup_end_state_dst().get_U_fn()
+
+        def U_fn(x):
+            return (1 - lamb) * U0_fn(x) + lamb * U1_fn(x)
```

Comment:
Maybe better for a TODO note and later revisit: Wondering how best to refactor to expose more parameters of `U0_fn`, `U1_fn`...

Reply:
willfix